### PR TITLE
fix: Remove global flag on regular expressions

### LINKED ISSUE #8175: DiscordAPIError: 404 on Client#fetchInvite() after Invite.INVITES_PATTERN#test()
### Which package is this bug report for?

discord.js

### Issue description

Client throws a DiscordAPIError: 404: Not Found error on a `Client.fetchInvite()` only after using it after a `Invites.INVITES_PATTERN.test(someString)`. Fetching before that indeed doesn't error, but for some reason that single regex check makes the whole thing stop working.

I'm marking this as High priority since it's a pretty breaking bug the fact that one cannot do an invite regex check and then fetch invites.

### Code sample

```typescript
// Censored actual invite code for obvious reasons

const { Client, Invite } = require('discord.js');

const inviteLink = 'https://discord.gg/someInviteCode';

async function run() {
    const client = new Client({ intents: ['GUILDS', 'GUILD_INVITES' ] });

    await client.login('TOKEN');
    
    const invite = await client.fetchInvite(inviteLink);
    console.log(invite);
    console.log('Check!');

    if (client.isReady() && Invite.INVITES_PATTERN.test(inviteLink)) {
        const inviteRegex = await client.fetchInvite(inviteLink);
        console.log(inviteRegex);
        console.log('Double check!');
    }
}

run();

/**
 * Console output:
 * 
 * Invite { guild: InviteGuild, code: 'someInviteCode', etc... }
 * Check!
 * 

D:\SRC\test\node_modules\discord.js\src\rest\RequestHandler.js:350
      throw new DiscordAPIError(data, res.status, request);
            ^

DiscordAPIError: 404: Not Found
    at RequestHandler.execute (d:\SRC\test\node_modules\discord.js\src\rest\RequestHandler.js:350:13)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async RequestHandler.push (d:\SRC\test\node_modules\discord.js\src\rest\RequestHandler.js:51:14)
    at async Client.fetchInvite (d:\SRC\test\node_modules\discord.js\src\client\Client.js:307:18)
    at async run (d:\SRC\test\index.js:15:29) {
  method: 'get',
  path: '/invites/https://discord.gg/someInviteCode?with_counts=true&with_expiration=true',
  code: 0,
  httpStatus: 404,
  requestData: { json: undefined, files: [] }
 */
```

### Package version

discord.js@13.8.1

### Node.js version

Node v16.14.0

### Operating system

Windows 10

### Priority this issue should have

High (immediate attention needed)

### Which partials do you have configured?

No Partials

### Which gateway intents are you subscribing to?

Guilds, GuildInvites

### I have tested this issue on a development release

_No response_

### PR DESCRIPTION
**Pull request by Jiralite**
**Please describe the changes this PR makes and why it should be merged:**
Resolves #8175 by removing the global flag on regular expressions. Some are now imported from discord-api-types and they all used named capturing groups now.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

**Review by SpaceEEC — APPROVED** [review]

File: packages/discord.js/src/structures/MessageMentions.js [hunk 1]
```diff
@@ -185,11 +190,14 @@ class MessageMentions {
   get channels() {
     if (this._channels) return this._channels;
     this._channels = new Collection();
+    const globalChannelsPatern = new RegExp(this.constructor.ChannelsPattern.source, 'g');
     let matches;
-    while ((matches = this.constructor.ChannelsPattern.exec(this._content)) !== null) {
-      const chan = this.client.channels.cache.get(matches[1]);
-      if (chan) this._channels.set(chan.id, chan);
+
+    while ((matches = globalChannelsPatern.exec(this._content)) !== null) {
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/MessageMentions.js:196]: Can't this just use matchAll and a for-of?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/MessageMentions.js (hunk 1 above)
Comment (kyranet) [reply packages/discord.js/src/structures/MessageMentions.js:196]: Discussed internally, this pattern yields 2x the performance compared to `string.matchAll(regexp)`.

File: packages/discord.js/src/structures/MessageMentions.js [hunk 2]
```diff
@@ -185,11 +190,14 @@ class MessageMentions {
   get channels() {
     if (this._channels) return this._channels;
     this._channels = new Collection();
+    const globalChannelsPatern = new RegExp(this.constructor.ChannelsPattern.source, 'g');
```
Comment (kyranet) [inline packages/discord.js/src/structures/MessageMentions.js:193]: Can we have this cached, please? I'm not a fan of recreating RegExp instances on the fly all the time. 😅

**Review by kyranet — APPROVED** [review]

File: packages/discord.js/src/structures/MessageMentions.js [hunk 3]
```diff
@@ -186,10 +199,12 @@ class MessageMentions {
...
     this._channels = new Collection();
     let matches;
-    while ((matches = this.constructor.ChannelsPattern.exec(this._content)) !== null) {
-      const chan = this.client.channels.cache.get(matches[1]);
-      if (chan) this._channels.set(chan.id, chan);
+
+    while ((matches = this.constructor.GlobalChannelsPattern.exec(this._content)) !== null) {
+      const channel = this.client.channels.cache.get(matches.groups.id);
+      if (channel) this._channels.set(channel.id, channel);
     }
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/MessageMentions.js:206]: Don't we have to reset lastIndex after?

Comment (imranbarbhuiya) [reply packages/discord.js/src/structures/MessageMentions.js:206]: while loop will stop after matching all matches. So once lastIndex becomes greater than input length, it'll become zero again

**Review by iCrawl — APPROVED** [review]

**Pull request merged**